swap/src/sqlite.rs

```diff
@@ -0,0 +1,257 @@
+DATABASE_URL=sqlite:tempdb cargo sqlx migrate run
+DATABASE_URL=sqlite:./swap/tempdb cargo sqlx prepare -- --bin swap
+```
+*/
+
+pub struct SqliteDatabase{
+    pool: Pool<Sqlite>
+}
+
+#[async_trait]
+impl Database for SqliteDatabase {
+    async fn open(path: PathBuf) -> Result<Self> where Self: std::marker::Sized  {
+        let pool = SqlitePool::connect(&*format!("sqlite:{}", &path.display()))
+            .await?;
+      Ok(Self { pool })
+    }
+
+    async fn insert_peer_id(&self, swap_id: Uuid, peer_id: PeerId) -> Result<()> {
+        let swap_id = swap_id.to_string();
+        let peer_id = peer_id.to_string();
+        let mut conn = self.pool.acquire().await?;
+        let swaps = sqlx::query!(
+        r#"
+        insert into peer_ids (
+            swap_id,
+            peer_id
+            ) values (?, ?);
+        "#,
+            swap_id, peer_id
+        ).execute(conn).await?;
+
+        Ok(())
+    }
+
+    async fn get_peer_id(&self, swap_id: Uuid) -> Result<PeerId> {
+        todo!()
+    }
+
+    async fn insert_monero_address(&self, swap_id: Uuid, address: Address) -> Result<()> {
+        let swap_id = swap_id.to_string();
+        let address = address.to_string();
+        let mut conn = self.pool.acquire().await?;
+        let swaps = sqlx::query!(
+        r#"
+        insert into monero_addresses (
+            swap_id,
+            address
+            ) values (?, ?);
+        "#,
+            swap_id, address
+        ).execute(conn).await?;
+
+        Ok(())
+    }
+
+    async fn get_monero_address(&self, swap_id: Uuid) -> Result<Address> {
+        todo!()
+    }
+
+    async fn insert_address(&self, peer_id: PeerId, address: Multiaddr) -> Result<()> {
+        let peer_id = peer_id.to_string();
+        let address = address.to_string();
+        let mut conn = self.pool.acquire().await?;
+        let swaps = sqlx::query!(
+        r#"
+        insert into monero_addresses (
+            peer_id,
+            address
+            ) values (?, ?);
+        "#,
+            peer_id, address
+        ).execute(conn).await?;
+
+        Ok(())
+    }
+
+    async fn get_addresses(&self, peer_id: PeerId) -> Result<Vec<Multiaddr>> {
+        todo!()
+    }
+
+    async fn insert_latest_state(&self, swap_id: Uuid, state: State) -> Result<()> {
+        let mut conn = self.pool.acquire().await?;
+        insert_latest_state(&mut conn, swap_id, state.into()).await
+    }
+
+    async fn get_state(&self, swap_id: Uuid) -> Result<State> {
+        let mut conn = self.pool.acquire().await?;
+        get_latest_state(&mut conn, swap_id).await
+    }
+
+    async fn all(&self) -> Result<Vec<(Uuid, State)>> {
+        let mut conn = self.pool.acquire().await?;
+        all_latest_swaps_iter(&mut conn).await
+    }
+
+    async fn unfinished(&self, unfinished: fn(State) -> bool) -> Result<Vec<(Uuid, State)>> {
+        Ok(self.all()
+            .await?
+            // todo: why is into iter returning an &State
```

Comment:
`into_iter` is not but `filter` only gives you a reference otherwise you would have to return the owned instance again in a positive case as you would be taking ownership.